### DISCUSSION TITLE
Some environment variables, defined in hyprland.conf, don't show up in printenv or echo output

### DISCUSSION THREAD

**Discussion by izmyname** [hyprwm/Hyprland#8204] [Q&A] [answered]
Some environment variables, defined in hyprland.conf like AQ_DRM_DEVICES and XDG_SESSION_DESKTOP, don't show up in printenv and echo output. For comparison, here's my hyprland.conf env variables.

```
#using integrated GPU for rendering
#https://wiki.hyprland.org/Configuring/Multi-GPU/
#lspci | \grep -E 'VGA|3D'
#ls -l /dev/dri/by-path
env = AQ_DRM_DEVICES,/dev/dri/card1

#XDG specs
env = XDG_SESSION_DESKTOP,Hyprland
env = XDG_CURRENT_DESKTOP,Hyprland
env = XDG_SESSION_TYPE,wayland

#cursor
env = XCURSOR_THEME,catppuccin-mocha-dark-cursors
env = XCURSOR_SIZE,24
env = HYPRCURSOR_THEME,catppuccin-mocha-dark
env = HYPRCURSOR_SIZE,24

#forcing everything to run on native wayland
env = GDK_BACKEND,wayland
env = SDL_VIDEODRIVER,wayland
env = CLUTTER_BACKEND,wayland
env = QT_QPA_PLATFORM,wayland
env = MOZ_ENABLE_WAYLAND,1
env = EGL_PLATFORM,wayland
env = ELECTRON_OZONE_PLATFORM_HINT,wayland 

#other vars
env = QT_QPA_PLATFORMTHEME,qt6ct
env = _JAVA_AWT_WM_NONREPARENTING,1
```

Here's printenv output

```
EDITOR=nvim
HOME=/home/arch-mage
LANG=en_GB.UTF-8
LOGNAME=arch-mage
MAIL=/var/spool/mail/arch-mage
PATH=/home/arch-mage/.local/bin:/usr/local/bin:/usr/bin
SHELL=/usr/bin/zsh
USER=arch-mage
VISUAL=nvim
XDG_DATA_DIRS=/home/arch-mage/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share/:/usr/share/
XDG_RUNTIME_DIR=/run/user/1000
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
DISPLAY=:0
HYPRLAND_INSTANCE_SIGNATURE=3bf6f78dad5f78dd8f4f519ceaa5a98671c90b14_1730920179_141353149
QT_QPA_PLATFORMTHEME=qt6ct
WAYLAND_DISPLAY=wayland-1
XDG_CURRENT_DESKTOP=Hyprland
MANAGERPID=1239
INVOCATION_ID=029525051dbf476980540d84d69b2a60
JOURNAL_STREAM=9:8832
SYSTEMD_EXEC_PID=1397
MEMORY_PRESSURE_WATCH=/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/foot-server.service/memory.pressure
MEMORY_PRESSURE_WRITE=c29tZSAyMDAwMDAgMjAwMDAwMAA=
TERM=foot-extra
COLORTERM=truecolor
PWD=/home/arch-mage
SHLVL=1
OLDPWD=/home/arch-mage
NNN_OPTS=ex
NNN_TERMINAL=foot -T preview-tui
NNN_ICONLOOKUP=1
NNN_BMS=c:~/.config/;o:~/.local/opt;l:~/.local/share;
NNN_PLUG=m:mtpmount;p:preview-tui;f:fzopen;M:nmount;S:suedit;u:getplugs;g:gpge;G:gpgd;i:-!|mediainfo "$nnn";l:-!bat --paging=always "$nnn"*;x:!&gio launch "$nnn";v:-!&swayimg -g "$nnn";s:-!dua "$nnn";w:-!&mpv "$nnn";a:autojump;
NNN_TRASH=1
NNN_COLORS=#04020301;4231
NNN_FCOLORS=030304020705050801060301
NNN_ARCHIVE=\.(7z|a|ace|alz|arc|arj|bz|bz2|cab|cpio|deb|gz|jar|lha|lz|lzh|lzma|lzo|rar|rpm|rz|t7z|tar|tbz|tbz2|tgz|tlz|txz|tZ|tzo|war|xpi|xz|Z|zip)$
NNN_FIFO=/tmp/nnn.fifo
RCLONE_PROGRESS=true
RCLONE_HUMAN_READABLE=true
DIFFPROG=nvim -d
PYTHONSTARTUP=/home/arch-mage/.config/python/pythonrc
_ZO_ECHO=1
_ZO_RESOLVE_SYMLINKS=1
_=/usr/bin/printenv

```
e.g  ` echo  $AQ_DRM_DEVICES` returns an empty string.

It's worth to mention, that:

1. This is a regression, as variables used to show up in the output on 0.43 and earlier.
2.  Variables do their thing, despite their absence in the output.
3.  The method of starting Hyprland (systemd service or by typing Hyprland in tty) doesn't matter.

**Comment by vaxerski** [comment]
uhhh... odd. Environment should be passed through with `execl`... Maybe I am missing something in posix

**Comment by izmyname** [comment]
Reopen this one, too, for further references or if it helps to solve some issues caused by it. This isn't really a problem, though, as variables still work.

**Comment by izmyname** [comment]
**UPD:** For uwsm users - just add `exec-once = uwsm finalize` to export those variables into environment. If some vars aren't exported (eg AQ_DRM_DEVICES, QT_QPA_PLATFORM) - just add them to the command as an argument (e.g, `exec-once = uwsm finalize AQ_DRM_DEVICES QT_QPA_PLATFORM`)

**Reply by autofool** [reply, marked as answer]
`uwsm` users should use systemd environment for variables. systemd-environemnt reads all files from $HOME/.config/environment.d/ on login. e.g.
```
/home/user/.config/environment.d/rust.conf
-----------------------------------------------------
RUST_HOME=$HOME/.config/rust
CARGO_HOME=$RUST_HOME/cargo
RUSTUP_HOME=$RUST_HOME/rustup
```

Plus, it expands the variables defined previously. For example, in hyprland if I `echo $CARGO_HOME` then the output will be `$RUST_HOME/cargo` while it should be `/home/user/.config/rust/cargo`.

**Reply by izmyname** [reply]
Both options seem to work. IMHO, `hyprland.conf` is good for graphical session -related vars (like electron, cursors, theming) and environment.d - for everything else.

Although, I use .zshrc for non gui-related vars.

**Reply by kai-tub** [reply]
I am using NixOS with home-manager and after updating today, I had to export the following PATHs via `environment.d` (`systemd.user.sessionVariables`):
```
"PATH" = "$PATH:/run/current-system/sw/bin/:/etc/profiles/per-user/<user>/bin/";
``` 
for fuzzel or any dispatchers to find the binaries.

**Reply by izmyname** [reply]
A small update - just transferred all my non-hypr vars from zshrc and hyprland.conf to environment.d, and it is, for sure, a way superior method to manage them.

**Reply by izmyname** [reply]
**upd:** There's a downside, though. In order to apply changes, after changing a variable's value, one must reload the whole session.

**Reply by izmyname** [reply]
Overall, to sum it up.

~/.config/environment.d - for setting vars for all sessions (like adding dirs to $PATH and the examples in the first comment).

~/.config/uwsm/env - setting vars for compositors, started with uwsm. ( QT_QPA_PLATFORM_THEME and GTK_THEME are ideal examples here - GNOME/KDE handle them in their own way, but compositors - don't)

~/.config/uwsm/env-compositorname - for compositor-specific vars (an ideal example - AQ_DRM_DEVICES inside ~/.config/uwsm/env-hyprland)

shell config - for cli apps, configured exclusively by setting vars (rclone, nnn, zoxide - anything, meant to be reloaded w/o logout)

hyprland.conf - for cursors, as uwsm exports them, but they can be configured inside ~/.config/uwsm/env and ~/.config/uwsm/env-hyprland, as well, I guess.

**Reply by Vladimir-csp** [reply]
> **upd:** There's a downside, though. In order to apply changes, after changing a variable's value, one must reload the whole session.

Not exactly, you can do `systemctl --user set-environment VAR=new_value` and then just restart or re-run the units that need the new value. If you run your apps as services, you can restart them like any other service unit (systemctl or uuctl). If you use scopes, then ~~you'll have to just run them again~~ you have to re-run them from somewhere with the new environment.

**Reply by izmyname** [reply]
Btw, is this okay to export per-user locale vars (LANG, LC_TIME) in `.config/environment.d/*.conf`? I used to use .`config/locale.conf`, as per [arch wiki](https://wiki.archlinux.org/title/Locale#Overriding_system_locale_per_user_session) instruction, but this method no longer is working for me.

**Reply by Vladimir-csp** [reply]
> Btw, is this okay to export per-user locale vars (LANG, LC_TIME) in .config/environment.d/*.conf?

I see no problem with that.

> If you use scopes, then you'll have to just run them again.

Correction: scopes inherit environment they were launched from, so only services can be restarted with updated environment.

**Reply by izmyname** [reply]
It turns out, `environment.d` variables aren't sourced without starting a graphical session, so all non-gui-related stuff should be exported in a shell config, e.g

`.zshenv/.profile/whatever `- $PATH

`environment.d/*.conf` - $BROWSER, maybe $TERMINAL - applies to all graphical sessions

`uwsm/env` -$XCURSOR, $QT_QPA_PLATFORM(THEME) - for uwsm-managed sessions, only.

`uwsm/env-hyprland` -$HYPRCURSOR, $AQ_DRM_DEVICES - for uwsm-hyprland session, only

**Reply by Vladimir-csp** [reply]
Yes, I've corrected [description](https://github.com/Vladimir-csp/uwsm#4-environments-and-shell-profile) recently.
Also there is [going](https://github.com/Vladimir-csp/uwsm/commit/6325d8bb9ca76b90c226fac849a6a236378d662b) to be a back-merge of `uwsm start` context in the next release if I don't come up with something else for [this case](https://github.com/Vladimir-csp/uwsm/issues/72).

**Reply by autofool** [reply]
> It turns out, `environment.d` variables aren't sourced without starting a graphical session

It's not gui but systemd managed sessions. You can start shell with uwsm and it should read variables.

**Comment by izmyname** [comment]
Closing this, as I'm using uwsm, thus the problem with the vars in `hyprland.conf` irrelevant to me.
